File: java/src/com/company/bs.java
package com.company;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class bs {
    static int n,m;
    static int[] arr=new int[100000];
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }
        int start,end,mid,cnt,answer;
        start= Arrays.stream(arr).max().getAsInt();
        answer=end=(int)1e9;
        while(start<=end) {
            mid=(start+end)/2;
            cnt=0;
            int moneyOnMe;
            int i=0;
            while(i<n){
                moneyOnMe=mid;
                cnt++;
                while(i<n && arr[i]<=moneyOnMe){
                    moneyOnMe-=arr[i];
                    i++;
                }

            }

            if(cnt>m){
                start=mid+1;
            }
            else if (cnt <= m) {
                end=mid-1;
                answer=mid;
            }
        }
        System.out.println(answer);

        br.close();
    }
}

File: C++/step3/bs.cpp
//////////////2343
#include <iostream>
#include <algorithm>
using namespace std;
const int INF=1e9;
int arr[100000];
int n,m;
int main(void){
  cin>>n>>m;
  int start,end=0,mid,answer;
  for(int i=0;i<n;i++){
    cin>>arr[i];
    end+=arr[i];
  }
  answer=end;
  start=*max_element(arr,arr+n);
  
  while(start<=end){
    mid=(start+end)/2;
    int cnt=1,sum=0;
    for(int i=0;i<n;i++){
      if(sum+arr[i]>mid){
        cnt++;
        sum=0;
      }
      sum+=arr[i];
    }
    if(cnt>m){
      start=mid+1;
    }
    else{
      answer=mid;
      end=mid-1;
    }
  }
  cout<<answer<<"\n";
  
  return 0;
}